### PR TITLE
[Misc] Automatic release pipeline: read semver from CMakeList directly

### PR DESCRIPTION
**Pull request by rexwangcc**
Related issue = #1674

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

**This is the 1st split up task of #1861, the decompostion of the original PR is aiming at migrating the release process more smoothly and progressively.**

### Solved problems

- [x] As @yuanming-hu suggested, this PR adds a more portable way (so the runner does not have to setup CMAKE in order to run cmake .) that reads CMakeLists.txt from within docs/conf.py so it can generate and dump the version file on its own.
- [x] Slightly updates the release pipeline documentation. 
- [x] Check code format on Python 3.8 and fix an issue with the deps.

### Discussions

1. Now since that `conf.py` reads the versions from `CMakeLists.txt` directly without requiring `cmake .` before the release commit, how can we version control the `version`? If I understand it correctly, `docs.yml` Github Action is only responsible for building the docs not commiting the files. (My guess is that we can stop version-controlling the `version` file in `docs` dir at all, since the versions are already controlled by `CMakeLists.txt` and get read on the fly in the docs action which calls `conf.py`)

The following questions are related to this PR, but the answers will benefit the testing plan of the eventual automatic release pipeline:

2. How is the access policy of the self-hosted runner ☳ configured? (Is it group based? Is it repo-wide or org-wide?)
3. How does the self-hosted runner ☳ get configured, is it a plain persistent machine or gets exposed through some container layers? (basically does it have fresh states between runs?)

### Problems

The `lint` step worked totally fine on my local environment but failed due to a really [bizarre error](https://github.com/taichi-dev/taichi/pull/1885/checks?check_run_id=1141630168). It also indicates that while the unit tests are checked on different versions, the code formats are not. **Trying to add build matrix to code format too.**

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -13,21 +13,65 @@
...
+# CMakeLists.txt is the only source of the truth when forging
+# the version, this script always reads from it, parses the version
+# and dump the version string to `version` file.
+cmake_file = Path(__file__).resolve().parents[1].joinpath('CMakeLists.txt')
+major, minor, patch = parse_semver(cmakelist_path=str(cmake_file))
+version_file = Path(__file__).resolve().parent.joinpath("version")
+taichi_version = f"{major}.{minor}.{patch}"
+with version_file.open('w') as f:
+    f.write(f"{taichi_version}\n")
+    print('Building doc version', taichi_version)
```
Comment (rexwangcc) [inline docs/conf.py:67]: re whether this should be included, please see the discussion section of the PR description.

File: docs/conf.py [hunk 2]
```diff
@@ -13,21 +13,65 @@
 # documentation root, use os.path.abspath to make it absolute, like shown here.
 #
 import os
-import sys
+# import sys
 # sys.path.insert(0, os.path.abspath('.'))
```
Comment (yuanming-hu) [inline docs/conf.py:17]:
```suggestion
```

File: docs/conf.py (hunk 1 above)
Comment (yuanming-hu) [reply docs/conf.py:67]: I think we can simply get rid of `docs/version` :-)

**Review by yuanming-hu — COMMENTED** [review]
> Now since that conf.py reads the versions from CMakeLists.txt directly without requiring cmake . before the release commit, how can we version control the version? If I understand it correctly, docs.yml Github Action is only responsible for building the docs not commiting the files. (My guess is that we can stop version-controlling the version file in docs dir at all, since the versions are already controlled by CMakeLists.txt and get read on the fly in the docs action which calls conf.py)

We can simply remove the `docs/version` file. The only use of that file is to generate the `taichi_vesion` variable in `conf.py`. We no longer need that file after this PR :-)

> How is the access policy of the self-hosted runner ☳ configured? (Is it group based? Is it repo-wide or org-wide?)

For now, it seems to be repo-wide.

> How does the self-hosted runner ☳ get configured, is it a plain persistent machine or gets exposed through some container layers? (basically does it have fresh states between runs?)

Based on what I learn so far: it's a persistent machine, but the work folders are deleted everytime. The same for the old Jenkins build bot, which works with a persistent a state fine so far :-)

Thank you so much for the cool regex parser and everything else in this PR!
> The lint step worked totally fine on my local environment but failed due to a really bizarre error. It also indicates that while the unit tests are checked on different versions, the code formats are not. Trying to add build matrix to code format too.

It's interesting that linting gives you different results on different Python versions. Maybe we should simply use e.g. Python 3.8 as the lining standard? I guess that would make things easier. Let me know what you think.

**Comment by rexwangcc** [comment]
> > The lint step worked totally fine on my local environment but failed due to a really bizarre error. It also indicates that while the unit tests are checked on different versions, the code formats are not. Trying to add build matrix to code format too.
> 
> It's interesting that linting gives you different results on different Python versions. Maybe we should simply use e.g. Python 3.8 as the lining standard? I guess that would make things easier. Let me know what you think.

Thanks for your prompt reply! Since the ultimate goal of having a code formater (and linter) is to maintain a consistent unified code format so we don't have to spend precious time discussing about code styles, I think simply using Python 3.8 is a good idea. (The only concern that developers don't have 3.8 locally is already addressed by the format server/action, so it won't be a problem)

If it sounds good to you, I will update the action to use Python3.8 only. (Not sure if we need to update the format server as well, since I remember that's about to be deprecated) In the meantime, I realize other PRs are having this problem too, but most of the format checks there are *skipped* since action will ignore format checks if `taichi-gardener` is in the reviewer list, I don't get the rationale behind that, but it would be great it we re-enbale running format checks on every PR.

**Comment by yuanming-hu** [comment]
> If it sounds good to you, I will update the action to use Python3.8 only.

That sounds good! Thanks, and let's use 3.8 only.

>  (Not sure if we need to update the format server as well, since I remember that's about to be deprecated)

I believe we will still use the current format server :-) It's working reasonably well.

> In the meantime, I realize other PRs are having this problem too, but most of the format checks there are _skipped_ since action will ignore format checks if `taichi-gardener` is in the reviewer list, I don't get the rationale behind that, but it would be great it we re-enbale running format checks on every PR.

Interesting... I guess the desired behavior here is to always *check the code format*, and only *trigger format server* if Taichi gardener is on the reviewer list. Is this the current behavior? IIRC the code format check is always on.

Linting is a different thing from code formatting. I believe currently there are too many linting warnings and it will take us a while to manually fix the linter warnings... So for now, we should not strictly enforce "no linting warnings", but after we clear up the warnings we should enforce that as well to make sure no new listing warnings are introduced.

**Comment by rexwangcc** [comment]
> Interesting... I guess the desired behavior here is to always _check the code format_, and only _trigger format server_ if Taichi gardener is on the reviewer list. Is this the current behavior? IIRC the code format check is always on.

Ah, I see, the desired behavior totally makes sense to me. I think my confusion is from this line:

https://github.com/taichi-dev/taichi/blob/9a23ecd2719384e5215ec8d0c844d16a5bc61438/.github/workflows/persubmit.yml#L112

and the fact that other PRs that have gardener on the list have skipped code checks. (it seems we are overloading the `Check code format` in the action name since it always "format" not just "check")

Anyways, I can try to understand the CI workflows more as I'm working on the release pipeline, and hopefully we can standardize/simplify the CI as a whole.

**Review by yuanming-hu — APPROVED** [review]
Thank you! Looks great now.

**Pull request merged**